winpr/libwinpr/utils/collections/HashTable.c

```diff
@@ -714,19 +799,58 @@ void HashTable_Free(wHashTable* table)
 			{
 				nextPair = pair->next;
 
-				if (table->keyFree)
-					table->keyFree(pair->key);
-
-				if (table->valueFree)
-					table->valueFree(pair->value);
-
-				free(pair);
+				disposePair(table, pair);
 				pair = nextPair;
 			}
 		}
-
-		DeleteCriticalSection(&(table->lock));
 		free(table->bucketArray);
-		free(table);
 	}
+	DeleteCriticalSection(&(table->lock));
+
+	free(table);
+}
+
+wObject* HashTable_KeyObject(wHashTable* table)
+{
+	WINPR_ASSERT(table);
+	return &table->key;
+}
+
+wObject* HashTable_ValueObject(wHashTable* table)
+{
+	WINPR_ASSERT(table);
+	return &table->value;
+}
+
+BOOL HashTable_SetHashFunction(wHashTable* table, HASH_TABLE_HASH_FN fn)
```

Comment:
Then `HashTable_SetHashFunction` shall probably return void as it can only assert

Reply:
True. 
The limit between input parameters checks and API violation that gives an assert is tight. We assert on `table == NULL` but we return FALSE on NULL `fn`. Should it be also an assert ?